rules/memory.go

```diff
@@ -127,32 +129,86 @@ func (m *MemoryStore) indexComponent(component components.Component) (set.Set[st
+		// iterates over the properties the parameters stored
+		// in the map are populated.
+		paramMap := make(map[string]extensions.Parameter)
+
 		for prop := range propSet {
-			switch prop.Name {
+			var propName string
+			var propSuffix string
+
+			// Matches any parameter properties that have a numerical suffix
+			re := regexp.MustCompile(`^Parameter_.*\d+$`)
+			// If the property name contains "Parameter" and has a numerical
+			// suffix then extract the suffix as the key for the map.
+			// Otherwise default to using "0" as the key (meaning there is
+			// no numerical suffix and only one parameter contained
+			// in the properties).
+			if re.MatchString(prop.Name) {
+				// Split the property name to handle properties that have
+				// a numerical suffix.  e.g Parameter_Id_1
+				propNameParts := strings.Split(prop.Name, "_")
+				propPrefix := propNameParts[:len(propNameParts)-1]
+				propName = strings.Join(propPrefix, "_")
+				propSuffix = propNameParts[len(propNameParts)-1]
+			} else {
+				propName = prop.Name
+				propSuffix = "0" // Used if property does not have numerical suffix
+			}
+
+			switch propName {
 			case extensions.RuleIdProp:
 				ruleSet.Rule.ID = prop.Value
 			case extensions.RuleDescriptionProp:
 				ruleSet.Rule.Description = prop.Value
+			case extensions.CheckIdProp:
+				placeholderCheck.ID = prop.Value
+			case extensions.CheckDescriptionProp:
+				placeholderCheck.Description = prop.Value
 			case extensions.ParameterIdProp:
-				if ruleSet.Rule.Parameter == nil {
-					ruleSet.Rule.Parameter = &extensions.Parameter{}
+				p, ok := paramMap[propSuffix]
+				if !ok {
+					paramMap[propSuffix] = extensions.Parameter{
+						ID: prop.Value,
+					}
+				} else {
+					p.ID = prop.Value
+					paramMap[propSuffix] = p
 				}
-				ruleSet.Rule.Parameter.ID = prop.Value
 			case extensions.ParameterDescriptionProp:
-				if ruleSet.Rule.Parameter == nil {
-					ruleSet.Rule.Parameter = &extensions.Parameter{}
+				p, ok := paramMap[propSuffix]
+				if !ok {
+					paramMap[propSuffix] = extensions.Parameter{
+						Description: prop.Value,
+					}
+				} else {
+					p.Description = prop.Value
+					paramMap[propSuffix] = p
 				}
-				ruleSet.Rule.Parameter.Description = prop.Value
-
 			case extensions.ParameterDefaultProp:
-				if ruleSet.Rule.Parameter == nil {
-					ruleSet.Rule.Parameter = &extensions.Parameter{}
+				p, ok := paramMap[propSuffix]
+
+				if !ok {
+					paramMap[propSuffix] = extensions.Parameter{
+						Value: prop.Value,
+					}
+				} else {
+					p.Value = prop.Value
+					paramMap[propSuffix] = p
 				}
-				ruleSet.Rule.Parameter.Value = prop.Value
-			case extensions.CheckIdProp:
-				placeholderCheck.ID = prop.Value
-			case extensions.CheckDescriptionProp:
-				placeholderCheck.Description = prop.Value
+			}
+		}
+
+		// Add any parameters that were extract from the
+		// properties to the rule
+		if len(paramMap) > 0 {
+			ruleParams := make([]extensions.Parameter, 0)
```

Comment:
This can avoid repeat re-allocations between you know the size upfront.
```suggestion
			ruleParams := make([]extensions.Parameter, len(paramMap))
```